src/gui/symbology-ng/qgssymbollayerv2widget.cpp

```diff
@@ -45,6 +45,28 @@
 #include <QStandardItemModel>
 #include <QSvgRenderer>
 #include <QMessageBox>
+#include <QTextDocument>
+
+void QgsSymbolLayerV2Widget::registerDataDefinedButton( QgsDataDefinedButton * button, const QString & propertyName, QgsDataDefinedButton::DataType type, const QString & description )
+{
+  QgsDataDefined dd( symbolLayer()->dataDefinedProperty( propertyName ) );
+  button->init( mVectorLayer, &dd, type, Qt::escape( description ) );
+  mDataDefinedPropertyButtons.insert( propertyName, button );
+  connect( button, SIGNAL( dataDefinedChanged( const QString& ) ), this, SLOT( updateDataDefinedProperty() ) );
+  connect( button, SIGNAL( dataDefinedActivated( bool ) ), this, SLOT( updateDataDefinedProperty() ) );
+}
+
+void QgsSymbolLayerV2Widget::updateDataDefinedProperty()
+{
+  symbolLayer()->removeDataDefinedProperties();
+  for ( QMap< QString, QgsDataDefinedButton* >::const_iterator it
+        = mDataDefinedPropertyButtons.begin(); it != mDataDefinedPropertyButtons.end(); ++it )
+  {
+    if ( it.value()->isActive() )
+      symbolLayer()->setDataDefinedProperty( it.key(), it.value()->currentDefinition() );
+  }
+  emit changed();
+}
```

Comment:
@vmora Here's a simpler approach: 
```C++
void QgsSymbolLayerV2Widget::registerDataDefinedButton( QgsDataDefinedButton * button, const QString & propertyName, QgsDataDefinedButton::DataType type, const QString & description )
{
  QgsDataDefined dd( symbolLayer()->dataDefinedProperty( propertyName ) );
  button->init( mVectorLayer, &dd, type, Qt::escape( description ) );
  button->setProperty( "propertyName", propertyName );
  connect( button, SIGNAL( dataDefinedChanged( const QString& ) ), this, SLOT( updateDataDefinedProperty() ) );
  connect( button, SIGNAL( dataDefinedActivated( bool ) ), this, SLOT( updateDataDefinedProperty() ) );
}

void QgsSymbolLayerV2Widget::updateDataDefinedProperty()
{
  QgsDataDefinedButton* button = qobject_cast<QgsDataDefinedButton*>( sender() );
  QString propertyName = button->property( "propertyName" ).toString();
  if ( button->isActive() )
  {
    symbolLayer()->setDataDefinedProperty( propertyName, button->currentDefinition() );
  }
  else
  {
    symbolLayer()->removeDataDefinedProperty( propertyName );
  }
  emit changed();
}
```

This avoids the need to remove all properties and rebuild them, and means that mDataDefinedPropertyButtons is no longer required.